docs/features/logforwarding/parse-json-structured-logs.adoc

```diff
@@ -0,0 +1,152 @@
+----
+
+
+
+
+=== Output type: Elasticsearch
+
+With Elasticsearch, JSON records with different formats *must* go to different indices, otherwise type conflicts and cardinality problems can occur.
+This is a *required* part of the configuration. The elasticsearch output must be configured with a *"structured type"* that is used to construct an index name.
+
+*  `structuredTypeKey`: (string, optional) Use the value of this meta-data key (if present and non-empty) as the structured type.
+*  `structuredTypeName`: (string, optional) Name to use for the structured type, unless _"structuredTypeKey"_ is set, and the value is present.
+
+==== Notes:
+* *At least one* of the structuredType options must be set when forwarding to an Elasticsearch type output
+* The Elasticsearch _index_ for structured records is formed by prepending "app-" to the structured type and appending "-write".
+* Unstructured records are not sent to the structured index, they are indexed as usual in application, infrastructure or audit indices.
+* If `structureTypeKey` is not found in the log entry, and no `structuredTypeName` is specified as a fallback, forward an _unstructured_ record without a `structured` field.
+
+
+==== Example
+
+For elasticsearch outputs, we must separate logs with different formats into different indices.
+Lets assume we have:
+
+* Applications logging in two structured JSON formats called "apache" and "google".
+* Each of our application pods have been labelled with the appropriate `logFormat=apache` or `logFormat=google`
+
+With the following forwarder configuration:
+
+.cluster-log-forwarder.yaml
+[source,yaml]
+----
+kind: ClusterLogForwarder
+apiVersion: logging.openshift.io/v1
+metadata:
+  name: instance
+  namespace: openshift-logging
+spec:
+  outputs:
+    - name: default
+      type: elasticsearch
+      elasticsearch:
+        structuredTypeKey: kubernetes.labels.logFormat
+  pipelines:
+    - name: my-formatted-logs
+      outputRefs:
+        - default
+      inputRefs:
+        - application
+      parse: json
+----
+. After parsing into _structured_, these log record will go to the index `app-apache-write`:
++
+[source,json]
+----
+{
+  "structured":{"name":"fred","home":"bedrock"},
+  "kubernetes":{"labels":{"logFormat": "apache", ...}}
+}
+----
+
+. This _structured_ log record will go to the index `app-google-write`:
++
+[source,json]
+----
+{
+  "structured":{"name":"wilma","home":"bedrock"},
+  "kubernetes":{"labels":{"logFormat": "google", ...}}
+}
+----
+
+*Note*: In the example above, only _structured_ logs that contain a `logFormat` label go to the `google` or `apache` index.
+All others go to the default application index as _unstructured_ records, including:
+
+* Records with missing or empty `logFormat` label.
+* Records that could not be parsed as JSON,  _even if_ they have a `logFormat` label.
+
+==== Elasticsearch Implementation Details
+It is *important* not to overload elasticsearch with too many indices.
```

Comment:
This needs to be some sort of BIG FAT WARNING NOTE.  Customers should only do this if they fully understand the implications.  This is an advanced usage of the API (one I have always had serious concerns about) and they really need to understand  the shard explosion possibility

Reply:
agreed